### ISSUE TITLE
ENH: Add example or implement dissolve of contiguous or otherwise related geometries

### ISSUE THREAD

**Issue by brendan-ward** [geopandas/geopandas#1715]
An example was added to `networkx` (https://github.com/networkx/networkx/pull/4381) that may be more appropriate here, either via implementation in the API or as an example (potentially as a more advanced example, it is a little complex).  It certainly has room to be improved further.

This was also raised in a [comment](https://github.com/geopandas/geopandas/pull/1568#issuecomment-675020374) on #1568 

This could leverage #1674 to capitalize on vectorized spatial relations between two GeoSeries - which in this case would be the same GeoSeries against itself.  The important part is that it uses a spatial index followed by the predicate operation.

In early testing, the linked example appears faster than PySAL `Queen` contiguity.  

Ideally, this would permit the user to specify the spatial relationship (within reason).  The linked example uses `intersects`, but it could also be expanded to `touches`, `overlaps`, etc.

**Comment by ljwolf** [comment]
In my experience, the performance really depends on the predicate & data size. `predicate='intersects'` is very fast even in large data, so `pygeos` generally beats the `libpysal.weights.Queen` constructor. But, `predicate='touches'` is orders of magnitude slower, so `libpysal.weights.Rook` generally beats `pygeos`, even in moderately-sized data. The performance of `Queen` is always comparable to that of `Rook`, so it's usually pretty consistent internally... 

[See, as an example, making the contiguity graph for US counties](https://gist.github.com/ljwolf/252ce076aca6e41b181ef462a7848c25)

**Comment by martinfleis** [comment]
@ljwolf I have a WIP version of Queen somewhere which uses pygeos to get coordinates (instead of looping) which brings reasonable speedups esp. on large gdfs. I'll dig it out.

**Comment by brendan-ward** [comment]
@ljwolf thanks for sharing your findings and example on this!

Agreed that `intersects` is a fast predicate compared to the others.  Aside from differences in the complexity of the predicate tests, one thing that is driving the large difference in time between `intersects` and `touches` is the availability of fast prepared predicates within GEOS.  `pygeos` uses prepared predicates for geometries with intersecting bounding boxes.  `intersects` has a prepared predicate function in GEOS for polygons (should be as fast as possible).  `touches` falls back to an unprepared predicate internally within GEOS.

One workaround I've found is to use `query_bulk` with `predicate='intersects'`, inner join back to the original geometries, discard the self-intersections, then use the pairwise predicates for `touches`.  But that depends very much on the nature of the data returned from `intersects`.

It might be possible for us to [optimize this a bit further in pygeos](https://github.com/pygeos/pygeos/issues/264), where depending on the predicate we can first run a prepared intersection then the intended (unprepared) predicate.

Perhaps I don't fully understand Queen contiguity, but I didn't think that it was strictly equivalent to the `intersects` predicate?  It looks like the implementation of Queen is based on shared vertices, which is more strict than `intersects`.

**Comment by ljwolf** [comment]
> I didn't think that it was strictly equivalent to the intersects predicate?

Indeed! They are each slightly different. The contiguity graph constructors focus on shared vertices/edges on the boundary. It's possible that two geometries intersect, but don't share a vertex on their boundary. For example, overlapping polygons intersect, but are not Queen neighbors. 

For most well-formed data (like the counties I've shown or the NUTS1 regions at the networkx example), the two criteria are indeed identical though. This is because we can assume that data is planar (no overlaps) and nodes that can be shared are indeed shared. 

To get exact "queen" graphs using pygeos without this assumption, you'd need to focus on relate patterns where  the boundaries intersected in at lest 0D (for queen) or 1d (for Rook) and with disjoint interiors.

**Comment by brendan-ward** [comment]
I think in the case of dissolving contiguous geometries (this issue), we'd want to default to the predicates instead of Queen or Rook.  My primary use case of this is around geometries that are often not well-formed.

Perhaps we could allow input of a DE9-IM relate pattern as an optional parameter that overrides predicate, so that the user could specify more strict relationships where appropriate?  But that also wouldn't likely beat the performance of Queen or Rook with the way that relate pattern testing is implemented.

Another way could approach that would be to decouple the actual dissolve from the grouping used to drive it, in which case the current dissolve function would be fine.  In which case we're back to deciding if examples on how to define contiguity groups is sufficient, or if we want to add more to the geopandas API.  Given the variety of ways to define contiguity, and ability to leverage optional packages like libpysal + networkx for some of that, I'm not sure if we should add this to the API.

**Comment by martinfleis** [comment]
Probably related is this PR https://github.com/pysal/libpysal/pull/335 implementing libpysal's `fuzzy_contiguity` based on `query_bulk` exposing predicate. Therefore you can do the following to dissolve based on predicate (+ some buffer options)

```py
W = libpysal.weights.fuzzy_contiguity(gdf, predicate='touches')
gdf.dissolve(W.component_labels)
```

I would not add it to our API but it may be a nice set of examples.
